lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,247 @@
+require 'securerandom'
+require 'open3'
+require 'ffi'
+
+module PuppetX
+  module PowerShell
+    class PowerShellManager
+      extend Puppet::Util::Windows::String
+      extend FFI::Library
+
+      @@instances = {}
+
+      def self.instance(cmd)
+        @@instances[:cmd] ||= PowerShellManager.new(cmd)
+      end
+
+      def initialize(cmd)
+        @stdin, @stdout, @ps_process = Open3.popen2(cmd)
+
+        Puppet.debug "#{Time.now} #{cmd} is running as pid: #{@ps_process[:pid]}"
+
+        at_exit { exit }
+      end
+
+      def execute(powershell_code, timeout_ms = 300 * 1000)
+        output_ready_event_name =  "Global\\#{SecureRandom.uuid}"
+        output_ready_event = self.class.create_event(output_ready_event_name)
+
+        # always need a trailing newline to ensure PowerShell parses code
+        code = <<-CODE
+        $event = [Threading.EventWaitHandle]::OpenExisting("#{output_ready_event_name}")
+        if ($runspace -eq $null)
+        {
+          # CreateDefault2 requires PS3
+          if ([InitialSessionState].GetMethod('CreateDefault2'))
+          {
+            $sessionState = [InitialSessionState]::CreateDefault2()
+          }
+          else
+          {
+            $sessionState = [InitialSessionState]::CreateDefault()
+          }
+          $runspace = [RunspaceFactory]::CreateRunspace($sessionState)
+          $runspace.Open()
+        }
+
+        $powershell_code = @'
+#{powershell_code}
+'@
+        try
+        {
+          $file = [IO.Path]::GetTempFileName()
```

Comment:
This method of writing files is less secure than it can be.  At the very least, we should use `[IO.Path]::GetTempPath()` and [[IO.Path]::GetRandomFileName](https://msdn.microsoft.com/en-us/library/system.io.path.getrandomfilename\(v=vs.110\).aspx)

I'm testing locally to try and find a way to keep the file open and have PowerShell read it, but not having much luck.  Also, I think there's some potential for running into encoding issues by passing the string into PowerShell and having it write the file, so I wonder if we can use some additional options when writing the file on the Ruby side to influence creation - the Ruby implementation lives roughly at https://github.com/ruby/ruby/blob/eee616315d406e91a232c33434f5195d72578b65/win32/win32.c#L6041-L6210